src/app/services/views/ServiceDetailView.vue

```diff
@@ -41,17 +41,38 @@
 
         <TabsWidget
           v-else
-          :tabs="getTabs(data.serviceInsight)"
+          :tabs="getTabs(data)"
         >
           <template #overview>
-            <ServiceSummary
-              :service="data.serviceInsight"
-              :external-service="data.externalService"
+            <DataSource
+              v-if="data.serviceType === 'external'"
+              v-slot="{ data: externalService, isLoading: externalServiceIsLoading, error: externalServiceError }: ExternalServiceSource"
+              :src="`/meshes/${route.params.mesh}/external-services/${route.params.service}`"
+            >
+              <LoadingBlock v-if="externalServiceIsLoading" />
+
+              <ErrorBlock
+                v-else-if="externalServiceError"
+                :error="externalServiceError"
+              />
+
+              <EmptyBlock v-else-if="externalService === undefined" />
```

Comment:
Is that even needed? Seems like it can't happen with the above check no?

Reply:
It is, currently.

It’s not the ideal way of handling this. A better approach would be turning the `data` object into an object following the following type spec:

```ts
interface DataSourceLoaded<T> {
  status: 'loaded'
  value: T
}

interface DataSourceLoading<T> {
  status: 'loading'
  value: undefined
}

interface DataSourceError<T> {
  status: 'loading'
  value: ApiError
}
```

And defining `data` as a union of those interfaces.

Then, `LoadingBlock` would be used if `data.status === 'loading'` and `ErrorBlock` would be used if `data.status === 'error'` (at which point TS would know `value` can be used as an instance of `ApiError`).

`EmptyBlock` wouldn’t come into play here because after checking for the loading and error status, we would know `value` is of type `T` (here `ExternalService`). For views where that makes sense (and we don’t want to rely on something else like KTable’s own empty state), we can use some custom check and `EmptyBlock`.

Reply:
Alright, I’m changing this. I’m not a fan of conflating the data-carrying variable to also imply the loading state even though it happens to be enough.

Reply:
@johncowen also added more info about the motivation of this recently: https://github.com/kumahq/kuma-gui/pull/1284#discussion_r1298149625

Reply:
> Didn't @johncowen point out in another PR that up until now the loading state was handled as data === undefined? Seems like this would solve this ambiguity without introducing further types.

Yes definitely, kind of half catching up here so not entirely read right through this looked at this code 😅 . But I think we should keep things as simple as possible, we should just be implementing all the detail views using DataSource as it stands and as we did for the list views.

The codebase is generally looking a lot simpler, welcoming and approachable now we have added DataSource and inlined a lot of the simpler code. I want to avoid overcomplicating things where we don't need to.